tsdb/chunks/head_chunks.go

```diff
@@ -0,0 +1,786 @@
+	cs := map[int]io.Closer{}
+	defer func() {
+		if returnErr != nil {
+			var merr tsdb_errors.MultiError
+			merr.Add(returnErr)
+			merr.Add(closeAllFromMap(cs))
+			returnErr = merr.Err()
+		}
+	}()
+
+	files, err := listChunkFiles(cdm.dir.Name())
+	if err != nil {
+		return err
+	}
+	if cdm.pool == nil {
+		cdm.pool = chunkenc.NewPool()
+	}
+
+	chkFileIndices := make([]int, 0, len(files))
+	for seq, fn := range files {
+		f, err := fileutil.OpenMmapFile(fn)
+		if err != nil {
+			return errors.Wrap(err, "mmap files")
+		}
+		cs[seq] = f
+		bs[seq] = realByteSlice(f.Bytes())
+		chkFileIndices = append(chkFileIndices, seq)
+	}
+
+	cdm.mmappedChunkFiles = bs
+	cdm.closers = cs
+	cdm.size = 0
+
+	// Check for gaps in the files.
+	sort.Ints(chkFileIndices)
+	if len(chkFileIndices) == 0 {
+		return nil
+	}
+	lastSeq := chkFileIndices[0]
+	for _, seq := range chkFileIndices[1:] {
+		if seq != lastSeq+1 {
+			return errors.Errorf("found unsequential head chunk files %d and %d", lastSeq, seq)
+		}
+		lastSeq = seq
+	}
+
+	for i, b := range cdm.mmappedChunkFiles {
+		if b.Len() < HeadChunkFileHeaderSize {
+			return errors.Wrapf(errInvalidSize, "invalid head chunk file header in file %d", i)
+		}
+		// Verify magic number.
+		if m := binary.BigEndian.Uint32(b.Range(0, MagicChunksSize)); m != MagicHeadChunks {
+			return errors.Errorf("invalid magic number %x", m)
+		}
+
+		// Verify chunk format version.
+		if v := int(b.Range(MagicChunksSize, MagicChunksSize+ChunksFormatVersionSize)[0]); v != chunksFormatV1 {
+			return errors.Errorf("invalid chunk format version %d", v)
+		}
+
+		maxt := binary.BigEndian.Uint64(b.Range(HeaderMaxtOffset, HeaderMaxtOffset+MintMaxtSize))
+		if maxt == 0 {
+			// This is possible if Prometheus crashes and the maxt was unwritten to the
+			// last head chunk file. As a safe buffer, we set the maxt to mint + 1.5 times head chunk file time range.
+			f, err := os.OpenFile(files[i], os.O_WRONLY|os.O_CREATE, 0666)
+			if err != nil {
+				return err
+			}
+			defer func() {
+				var merr tsdb_errors.MultiError
+				merr.Add(returnErr)
+				merr.Add(f.Close())
+				returnErr = merr.Err()
+			}()
+			mint := binary.BigEndian.Uint64(b.Range(HeaderMintOffset, HeaderMintOffset+MintMaxtSize))
+			binary.BigEndian.PutUint64(cdm.writeBuf[:], mint+(uint64(cdm.maxFileMs)*3/2))
+			if _, err := f.WriteAt(cdm.writeBuf[:MintMaxtSize], HeaderMaxtOffset); err != nil {
+				return err
+			}
+		}
+
+		cdm.size += int64(b.Len())
+	}
+
+	return nil
+}
+
+func listChunkFiles(dir string) (map[int]string, error) {
+	files, err := ioutil.ReadDir(dir)
+	if err != nil {
+		return nil, err
+	}
+	res := map[int]string{}
+	for _, fi := range files {
+		seq, err := strconv.ParseUint(fi.Name(), 10, 64)
+		if err != nil {
+			continue
+		}
+		res[int(seq)] = filepath.Join(dir, fi.Name())
+	}
+	return res, nil
+}
+
+// WriteChunk writes the chunk to the disk.
+// The returned chunk ref is the reference from where the chunk encoding starts for the chunk.
```

Comment:
Without having a context it is unclear why we need maxt and mint here. I think it would be usefull to either explain here or in the head chunk readme or anywhere else.

Reply:
I will add it in the docs where we have the layout of the chunks written.